clang/lib/CodeGen/CGBuiltin.cpp

```diff
@@ -6088,6 +6088,9 @@ RValue CodeGenFunction::EmitBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
   StringRef Prefix =
       llvm::Triple::getArchTypePrefix(getTarget().getTriple().getArch());
   if (!Prefix.empty()) {
+    if (Prefix == "spv" &&
+        getTarget().getTriple().getOS() == llvm::Triple::OSType::AMDHSA)
+      Prefix = "amdgcn";
```

Comment:
What is this required for? I'm wondering why we'd need to reset the prefix here instead of updating the logic somewhere else.

Reply:
I don't quite see any other point to tweak the logic. We have some `amdgcn` builtins that get handled here, as opposed to via `EmitTargetBuiltinExpr`, and this is the narrowest scope for adjusting `Prefix`; the information that this is coming from AMDGCN flavoured SPIRV is lost past this point. Did you have something in mind regarding where to handle this as an alternative?

Reply:
So I was just wondering if it would make more sense to put this in `Triple::getArchTypePrefix(ArchType Kind)` because I wasn't sure if this logic is the expected return value there.

Reply:
I don't think we can, with the current signature of that interface / the current approach of keeping `spirv64` as the prefix, since we need at least the OS type as well to disambiguate. So it'd require either overloading that interface, or using a different prefix (this'd ripple elsewhere).

Reply:
If we want to allow this choice to be target-specific, we can just make the "IntrinsicPrefix" a property of the TargetInfo, instead of trying to derive it directly from the target triple.  The default constructor would derive it from the triple, but specific targets could override the field.

I'm more concerned this is going to prevent us from using builtins which shoud be usable; do we need to getIntrinsicForClangBuiltin() for both "spv" and "amdgcn"?

Reply:
The second bit is actually a good catch, I had assumed there's no SPIRV builtins, however looking at it it seems a few exist, so it appears we'll need both. I'll rework this part, cheers!

Reply:
I've fixed the `"spv"` exclusion. Circling back to the idea of moving the prefix into `TargetInfo`, that seems like a pretty nice improvement - I think we could also deal with scenarios like the one here where there's an union of target specific builtins (spv & amdgcn in this case) by having some sort of parent-child / main-aux relationship. It should probably be a separate PR since it's an independent piece of functionality (and I've not thought through it fully yet) - unless you disagree?

Reply:
We can look at it as a followup, sure; the code in its current form seems okay for now.